### PR TITLE
feat: Collect new statistics for the Contact Identification feature

### PR DESCRIPTION
**Pull request by matheusbsilva137**
## Proposed changes (including videos or screenshots)

 Added statistics related to the new **Contact Identification** feature:
 - `totalContacts`: Total number of contacts;
 - `totalUnknownContacts`: Total number of unknown contacts;
- `totalMergedContacts`: Total number of merged contacts (only contacts that have been deleted on merge are considered merged -- eg if one contact has other 2 similar ones to be merged in it, we'll count 2 merged contacts, NOT 3);
- `totalConflicts`: Total number of merge conflicts;
- `totalResolvedConflicts`: Total number of resolved conflicts;
- `totalBlockedContacts`: Total number of blocked contacts;
- `totalPartiallyBlockedContacts`: Total number of contacts partially blocked -- only contacts that have at least one blocked contact (but not all of them) are counted;
- `totalFullyBlockedContacts`: Total number of contacts fully blocked;
- `totalVerifiedContacts`: Total number of verified contacts;
- `avgChannelsPerContact`: Average number of channels per contact;
- `totalContactsWithoutChannels`: Number of contacts without channels;
- `totalImportedContacts`: Total number of imported contacts;
- `totalUpsellViews`: Total number of "Advanced Contact Management" Upsell CTA views;
- `totalUpsellClicks`: Total number of "Advanced Contact Management" Upsell CTA clicks;

## Further comments

For load test purposes, I populated my local DB with +2M contacts (2069065 contacts exactly) using [this filler repo](https://github.com/KevLehman/filler) which I extended in a fork so that the contacts collection could also be populated randomly considering all of its optional fields and arrays of random sizes. Using this DB, I ran each new query 10 times and calculated the average execution time (in milliseconds) and a 95% confidence interval for each of them (of course, with this we're making an assumption that the average execution time for each query follow a normal distribution):
 - `totalConflicts` and `avgChannelsPerContact` (they're calculated together): 3,444.1 ±75.577 (±2.19%);
 - `totalUnknownContacts`: 324.6 ±6.268 (±1.93%);
 - `totalBlockedContacts`: 589 ±18.433 (±3.13%);
 - `totalFullyBlockedContacts`: 2,743.6 ±41.011 (±1.49%);
 - `totalPartiallyBlockedContacts`: infered from `totalBlockedContacts` and `totalFullyBlockedContacts`;
- `totalVerifiedContacts`: 610.7 ±23.301 (±3.82%);
- `totalContactsWithoutChannels`: 819.9 ±23.373 (±2.85%);

All new queries used a index, except `totalContactsWithoutChannels`, `totalConflicts` and `avgChannelsPerContact` which all need to read all documents (so I believe they're all essentially a COLSCAN).

[SCI-78]

[SCI-78]: https://rocketchat.atlassian.net/browse/SCI-78?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by dougfabris — DISMISSED** [review]
approving on behalf of frontend!

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	] = await Promise.all([
+		LivechatContacts.countDocuments({}),
+		LivechatContacts.countDocuments({ unknown: true }),
+		LivechatContacts.countTotalAmountOfConflicts().toArray(),
+		LivechatContacts.countDocuments({ 'channels.blocked': true }),
+		LivechatContacts.countDocuments({
+			'channels.blocked': true,
+			'channels': { $not: { $elemMatch: { $or: [{ blocked: false }, { blocked: { $exists: false } }] } } },
+		}),
+		LivechatContacts.countDocuments({ unknown: false }),
```
[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts:25]: Actually, I believe we can't do that and I might be calculating verified contacts the wrong way
Verified contacts should be the ones that have any verified channel, not exactly those that aren't unknown. Please correct me if I'm  wrong, @pierre-lehnen-rc

File: apps/meteor/server/models/raw/LivechatContacts.ts [hunk 2]
```diff
@@ -217,4 +232,36 @@ export class LivechatContactsRaw extends BaseRaw<ILivechatContact> implements IL
...
+							$avg: { $size: { $cond: [{ $isArray: '$channels' }, '$channels', []] } },
+						},
+					},
+				},
+			],
+			{ allowDiskUse: true, readPreference: readSecondaryPreferred() },
+		);
+	}
+
+	countTotalAmountOfConflicts(): AggregationCursor<{ totalConflicts: number }> {
```
[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply apps/meteor/server/models/raw/LivechatContacts.ts]: Good idea, done!

File: apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		totalContacts,
+		totalUnknownContacts,
+		[{ totalConflicts, avgChannelsPerContact } = { totalConflicts: 0, avgChannelsPerContact: 0 }],
+		totalBlockedContacts,
+		totalFullyBlockedContacts,
+		totalVerifiedContacts,
+		totalContactsWithoutChannels,
+	] = await Promise.all([
+		LivechatContacts.estimatedDocumentCount(),
+		LivechatContacts.countDocuments({ unknown: true }),
```
Comment (sampaiodiego) [inline apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts:17]: pls move these queries to the model.. i.e.:

- `LivechatContacts.countUnknown()`
- `LivechatContacts.countKnown()`
- `LivechatContacts.countChannelsBlocked()`
- ...

**Review by sampaiodiego — CHANGES_REQUESTED** [review]

**Comment by sampaiodiego** [comment]
is it asking too much to create some data, perform some analysis of the queries and attach to the PR? I mean, I wonder if we should start being more cautions adding more metrics, specially when talking about omnichannel metrics as we have some HUGE datasets out there

File: apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tapiarafael) [reply apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts:24]: Yep. Contact unknown and Contact verified are two different things. We even can have a verified contact that is unknown.

File: apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts (hunk 3 above)
Comment (matheusbsilva137) [reply apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts]: Done!

File: apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply apps/meteor/app/statistics/server/lib/getContactVerificationStatistics.ts]: Fixed! I'm using the verified flag instead now (I had to add another index to cover this, though)
tks, @tapiarafael

**Review by MarcosSpessatto — DISMISSED** [review]

**Review by KevLehman — APPROVED** [review]

File: apps/meteor/server/models/raw/LivechatContacts.ts [hunk 4]
```diff
@@ -272,4 +292,46 @@ export class LivechatContactsRaw extends BaseRaw<ILivechatContact> implements IL
 
 		return this.countDocuments(filter);
 	}
+
+	countUnknown(): Promise<number> {
+		return this.countDocuments({ unknown: true });
```
Comment (sampaiodiego) [inline apps/meteor/server/models/raw/LivechatContacts.ts:297]: all this counts could be read from secondary as well

**Review by sampaiodiego — CHANGES_REQUESTED** [review]

**Review by sampaiodiego — APPROVED** [review]

**Review by dougfabris — APPROVED** [review]

**Review by KevLehman — APPROVED** [review]

**Pull request merged**